### PR TITLE
Resolves AB#9138 - BannerListener dying

### PR DESCRIPTION
**Pull request by sslaws**
# Fixes or Implements [AB#9138](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/9138)

* [ ] Enhancement
* [X] Bug
* [ ] Documentation

## Description

Increases logging and should fix an issue where the BannerListener exited early.  If the BannerListener dies early it will be restarted 10 seconds later.
Removed Startup and Program from CodeCoverage results for Unit tests.

**I've also removed the issue in Startup that prevents launch of the web app**

## Testing

* [ ] Unit Tests Updated
* [ ] Functional Tests Updated
* [X] Not Required

### Steps to Reproduce

One can manually reproduce by terminating the Postgres connection to the BannerListener and see that Banner Updates are no longer picked up.

### UI Changes

No

### Browsers Tested

* [ ] Chrome
* [ ] Safari
* [ ] Edge
* [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant.  Include any specialized deployment steps here.  

</br>
</br>
</br>
___
  
## Reviewer Guidelines

A reviewer should examine the PR if the Status above indicates the PR is **Ready**.
Once ready, the reviewer should read the information below as well as supporting information like the Azure ticket to ensure they have a good understanding of the request.

The reviewer should then:

1. Is the PR template complete and accurate?

    Confirm by reading all sections below.

1. Confirm the merge branch

    A normal feature/bug branch should merge into the Dev branch.  Exceptions for hotfix and sprint merges do occur and the target would then be Master.

1. Do the code commits have an appropriate comment and include linkage to the azure ticket by including a reference to "AB#nnnn"?

1. Review the Code Changes

    1. Do the changes match the description?

        The code commits should reflect the documented description and nothing more.  If additional changes are included the reviewer should call these out and request that the developer document the changes in Azure and in the description or remove the changed code.

    1. useful tips

        * Any obvious errors or omissions?
        * Does the code match the style of other similar work?
        * Is the code easy to understand?
        * Is the code well documented?
        * DB Migrations are appropriate?

If the Answer to any of the questions above is No, then a request for change should be the result of the review.

### CODE REVIEW

**Review by FuriousLlama — APPROVED** [review]

File: Apps/WebClient/src/Server/Startup.cs [hunk 1]
```diff
@@ -168,7 +170,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                         new SpaOptions { SourcePath = "ClientApp" },
                         npmScript: "serve",
                         port: 8585,
-                        regex: "Compiled successfully",
+                        regex: "Compiled ",
```
Comment (leobcgov) [inline Apps/WebClient/src/Server/Startup.cs:173]: should we leave the "successfully" here and wait for the fix?

File: Apps/WebClient/src/Server/Startup.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@ namespace HealthGateway.WebClient
 {
     using System;
     using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
```
Comment (leobcgov) [inline Apps/WebClient/src/Server/Startup.cs:21]: Is it ok to add this to the prod build?

File: Apps/WebClient/src/Server/Listeners/BannerListener.cs [hunk 3]
```diff
@@ -74,40 +75,52 @@ public class BannerListener : BackgroundService
...
+                        Thread.Sleep(SleepDuration);
+                    }
                 }
 
-                this.logger.LogInformation($"Exiting WaitChannelNotification: {Channel}");
-            }
-            catch (NpgsqlException e)
-            {
-                this.logger.LogError($"DB Error encountered in WaitChannelNotification: {Channel}\n{e.ToString()}");
+                this.logger.LogWarning($"DBChangeListener on {Channel} exiting...");
```
Comment (leobcgov) [inline Apps/WebClient/src/Server/Listeners/BannerListener.cs:117]: any reasons why we log warning when exception occurs here?

File: Apps/WebClient/src/Server/Listeners/BannerListener.cs [hunk 4]
```diff
@@ -74,40 +75,52 @@ public class BannerListener : BackgroundService
...
         protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
-            this.logger.LogDebug($"Registering Channel Notification on channel {Channel}");
+            this.logger.LogInformation("DBChangeListener is starting");
             stoppingToken.Register(() =>
-                this.logger.LogInformation($"DBChangeListener exiting..."));
-            try
+                this.logger.LogInformation($"DBChangeListener Shutdown as cancellation requested    "));
+            int attempts = 0;
+            while (!stoppingToken.IsCancellationRequested)
```
Comment (leobcgov) [inline Apps/WebClient/src/Server/Listeners/BannerListener.cs:82]: I have a feeling that this needs some improvements but not sure how. Should we invest a bit further to see if there is another way to do this without 2 whiles?

**Review by leobcgov — CHANGES_REQUESTED** [review]
need @sslaws to review my comments to make sure these changes

File: Apps/WebClient/src/Server/Startup.cs (hunk 1 above)
Comment (FuriousLlama) [reply Apps/WebClient/src/Server/Startup.cs:173]: not necessary, if it compiles it should work ok, even with warnings

File: Apps/WebClient/src/Server/Startup.cs (hunk 2 above)
Comment (sslaws) [reply Apps/WebClient/src/Server/Startup.cs:21]: Yes, lots of our code has it.

File: Apps/WebClient/src/Server/Listeners/BannerListener.cs (hunk 3 above)
Comment (sslaws) [reply Apps/WebClient/src/Server/Listeners/BannerListener.cs:117]: I was thinking that the outer loop is recovering the exception which means its no longer an error state but rather a warning which we should investigate.

File: Apps/WebClient/src/Server/Listeners/BannerListener.cs (hunk 4 above)
Comment (sslaws) [reply Apps/WebClient/src/Server/Listeners/BannerListener.cs:82]: Yeah when I first wrote it I thought it strange but the two loops are handling different things and just ensuring that we exit properly if a shutdown request comes in.  The outer loop is ensuring that we rebuild our listener and the inner loop is waiting on events from the DB.

**Review by leobcgov — APPROVED** [review]

**Review by sam-warren — APPROVED** [review]

**Pull request merged**